---
title: "CSAMA 2017: Remote data, remote computing"
author: Vince Carey
date: June 15, 2017
output: 
  ioslides_presentation:
    fig_height: 4.8
---

```{r setup, echo=FALSE, results="hide"}
suppressPackageStartupMessages({
library(dplyr)
library(bigrquery)
library(gwascat)
library(ldblock)
library(cgcR)
library(Rsamtools)
})
```

## Basic idea

We've made some assumptions in the basic curriculum

- local computing
    - you have control over the whole system
- local data
    - all the data are on your machine
- data in memory
- functions return results relatively quickly
- results are digestible in concise form

## Exceptions to these basic assumptions

- huge data
    - pan-cancer, TCGA, 1000 genomes and beyond
- algorithms with very large RAM requirements
    - alignments; fragment-level bias modeling for RNA-seq (alpine)
- analysis with very large numbers of tests
    - all SNPs vs all (transcripts, histone marks, ...)
    - long runtimes, intermittent failures, massive outputs
- can Bioconductor practices still be followed?

## What are the practices to be followed? (a selected few!)
 
- self-documenting data structures
    - integrate assay, sample-level, and experiment-level data
- functional object-oriented programming
    - `X[G, S]` is a faithful restriction/reorganization of `X`
and retains type/behavior of `X`
    - critical elements of metadata such as genome build are
bound tightly to entities whose interpretation require
knowledge of this
- all operations are evaluations of R functions
- R packages are fundamental for organizing, documenting, and testing
    - software, annotation, and data

## Example: GWAS catalog

```{r lkgw}
library(gwascat)
data(ebicat38)
ebicat38
```

## Beyond local computing 1: 1000 genomes in the cloud
```{r lk1k}
library(ldblock)
path(s3_1kg("22"))
# can use scanVcfHeader, 
#      readVcf(..., ScanVcfParam(which=[GRanges]))
```

## Beyond local computing 2: TCGA in BigTable, thanks to ISB/NCI
```
#package cgcR has isbApp(bq)
> getBQ
function () 
{
    library(dplyr)
    library(bigrquery)
    con <- DBI::dbConnect(dbi_driver(), project = "isb-cgc", 
        dataset = "tcga_201607_beta", billing = "...")
    con
}
```


## Beyond local computing 3: shinyapps.io

## Conclusions

- Bioconductor site includes considerable documentation on how to build bioc-enabled clusters in EC2
- New technical options are emerging, 
- Required: Curating the environments/strategies so that bioconductor principles (self-identifying, packaging with test disciplines, ...) are preserved as we work
